applications/MappingApplication/custom_utilities/mapping_matrix_builder.cpp

```diff
@@ -0,0 +1,68 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Philipp Bucher (https://github.com/philbucher)
+//
+
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "utilities/parallel_utilities.h"
+#include "mapping_matrix_builder.h"
+
+
+namespace Kratos
+{
+
+template<bool IsDistributed>
+void MappingMatrixBuilder<IsDistributed>::BuildMappingMatrix(
+    MapperLocalSystemPointerVector& rLocalSystems,
+    MappingMatrixPointerType& rpMappingMatrix,
+    InterfaceVectorPointerType& rpInterfaceVectorOrigin,
+    InterfaceVectorPointerType& rpInterfaceVectorDestination)
+{
+    GraphType matrix_graph;
+    GraphType vec_origin_graph;
+    GraphType vec_destination_graph;
+
+    MapperLocalSystem::EquationIdVectorType origin_ids;
+    MapperLocalSystem::EquationIdVectorType destination_ids;
+
+    for (auto& r_local_sys : rLocalSystems) {
+        r_local_sys->EquationIdVectors(origin_ids, destination_ids);
+        matrix_graph.AddEntries(origin_ids, destination_ids);
+        vec_origin_graph.AddEntries(origin_ids);
+        vec_destination_graph.AddEntries(destination_ids);
+    }
+
+    MappingMatrixPointerType p_matrix(Kratos::make_unique<MappingMatrixType>(matrix_graph));
+    rpMappingMatrix.swap(p_matrix);
+
+    InterfaceVectorPointerType p_vec_o(Kratos::make_unique<InterfaceVectorType>(vec_origin_graph));
+    rpInterfaceVectorOrigin.swap(p_vec_o);
+
+    InterfaceVectorPointerType p_vec_d(Kratos::make_unique<InterfaceVectorType>(vec_destination_graph));
+    rpInterfaceVectorDestination.swap(p_vec_d);
```

Comment:
@RiccardoRossi this is how I would do it
Can you please take a look and let me know?

Initial question, why does the DistributedGraph need a size? => I don't know this (I don't even understand what it represents)

Reply:
do you need to assemble into InterfaceVectorOrigin and Destionation, or you just need them as vectors?

Reply:
2 comments about this:

1 - note that the vector does NOT take ownership of the graph

other than that if you know the number of local rows for the csr matrix and vectors than you should use the SparseContiguousRowGraph instead which would allow you a parallel construction.

If you don't have that information, i would rewrite your code as

~~~cpp
    GraphType matrix_graph; //here i would use the SparseContiguousRowGraph to be able to construct it in parallel, but it would need the size

    MapperLocalSystem::EquationIdVectorType origin_ids;
    MapperLocalSystem::EquationIdVectorType destination_ids;

    for (auto& r_local_sys : rLocalSystems) {
        r_local_sys->EquationIdVectors(origin_ids, destination_ids);
        matrix_graph.AddEntries(origin_ids, destination_ids);
    }

    MappingMatrixPointerType p_matrix(Kratos::make_unique<MappingMatrixType>(matrix_graph));
    rpMappingMatrix.swap(p_matrix);

    //i am separating this in two loops to assume that you don't know the local size, otherwise you could just use it here from the start
    InterfaceVectorPointerType vec_destination_graph(rpMappingMatrix.GetRowNumbering() );
    InterfaceVectorPointerType vec_origin_graph(rpMappingMatrix.GetColNumbering() ); //one is Col the other is Row but i am not 100% sure of what in your implementation

    for (auto& r_local_sys : rLocalSystems) {
        r_local_sys->EquationIdVectors(origin_ids, destination_ids);
        rpInterfaceVectorOrigin->AddEntries(origin_ids); //note that here i am adding the entries directly to the vector, since you don't NEED a graph for that (if you have you can use it)
        rpInterfaceVectorDestination->AddEntries(destination_ids);
    }
    
    rpInterfaceVectorOrigin.swap(p_vec_o);
    rpInterfaceVectorDestination.swap(p_vec_d);
~~~

also note that adding the entries to the vector is only needed if you have to assemble onto it, if you just need it for SpMV like operations than it is not needed